Fix typing for `select`

ErichDonGubler opened the pull request:
**Connections**

- ~~Depends on #7602.~~
- Resolves #7538.

**Testing**

Tests have been added, as well as a test run against [`webgpu:shader,execution,expression,call,builtin,select:*`](https://gpuweb.github.io/cts/standalone/?q=webgpu:shader,execution,expression,call,builtin,select:*) in a locally modified Firefox.

**Squash or Rebase?**

rebase plz

**Checklist**

- [x] If this contains user-facing changes, add a `CHANGELOG.md` entry. <!-- See instructions at the top of `CHANGELOG.md`. -->

ErichDonGubler committed:
fix(naga): properly impl. auto. type conv. for `select`

jimblandy commented:
I hate to say it, but you might compare this approach with just using stuff from `naga::front::wgsl::lower::conversion` directly.

We certainly prefer to use common logic over bespoke. But if you have to introduce a lot of bespoke code *just to use* the common logic, then we don't come out ahead. This is where I ended up with #7548, #7567, and #7573, unfortunately.

jimblandy commented:
Before we can allow abstract-typed arguments to select, we'll need to support it in the constant evaluator, so it might be best to start with a PR for that.

ErichDonGubler commented:
> I hate to say it, but you might compare this approach with just using stuff from `naga::front::wgsl::lower::conversion` directly.
> 
> We certainly prefer to use common logic over bespoke. But if you have to introduce a lot of bespoke code _just to use_ the common logic, then we don't come out ahead. This is where I ended up with #7548, #7567, and #7573, unfortunately.

I'm actually not sure that I share this opinion, with the progress I made today. Take a look, see what you think! The shared code doesn't look too bad IMO, and it feels pretty easy to turn around and share with other parts of the codebase in in its current state.

ErichDonGubler commented:
RE: status: I'm mostly at the point where I feel like I'm not quite understanding the intent behind `Debug` vs. `ToWgsl` between validation and the WGSL frontend. I think it _could_ make sense, but I need to internalize it better. I specifically resisted delving into them for the purposes of landing function database stuff, but now I need to keep things consistent with my own changes. 🙂

jimblandy commented:
For the WGSL front end, all that's needed is this:

```patch
modified   naga/src/front/wgsl/lower/mod.rs
@@ -2483,15 +2483,29 @@ impl<'source, 'temp> Lowerer<'source, 'temp> {
                         "select" => {
                             let mut args = ctx.prepare_args(arguments, 3, span);
 
-                            let reject = self.expression(args.next()?, ctx)?;
-                            let accept = self.expression(args.next()?, ctx)?;
+                            let mut values = [
+                                self.expression_for_abstract(args.next()?, ctx)?,
+                                self.expression_for_abstract(args.next()?, ctx)?,
+                            ];
+                            for &value in &values {
+                                ctx.grow_types(value)?;
+                            }
+                            if let Ok(consensus_scalar) =
+                                ctx.automatic_conversion_consensus(&values)
+                            {
+                                ctx.convert_slice_to_common_leaf_scalar(
+                                    &mut values,
+                                    consensus_scalar,
+                                )?;
+                            }
+
                             let condition = self.expression(args.next()?, ctx)?;
 
                             args.finish()?;
 
                             ir::Expression::Select {
-                                reject,
-                                accept,
+                                reject: values[0],
+                                accept: values[1],
                                 condition,
                             }
                         }
```

jimblandy commented:
> RE: status: I'm mostly at the point where I feel like I'm not quite understanding the intent behind `Debug` vs. `ToWgsl` between validation and the WGSL frontend. I think it _could_ make sense, but I need to internalize it better. I specifically resisted delving into them for the purposes of landing function database stuff, but now I need to keep things consistent with my own changes. 🙂

Could you file an issue for this, so the ideas don't get lost?

ErichDonGubler commented:
> For the WGSL front end, all that's needed is this:
> 
> ```diff
> modified   naga/src/front/wgsl/lower/mod.rs
> @@ -2483,15 +2483,29 @@ impl<'source, 'temp> Lowerer<'source, 'temp> {
>                          "select" => {
>                              let mut args = ctx.prepare_args(arguments, 3, span);
>  
> -                            let reject = self.expression(args.next()?, ctx)?;
> -                            let accept = self.expression(args.next()?, ctx)?;
> +                            let mut values = [
> +                                self.expression_for_abstract(args.next()?, ctx)?,
> +                                self.expression_for_abstract(args.next()?, ctx)?,
> +                            ];
> +                            for &value in &values {
> +                                ctx.grow_types(value)?;
> +                            }
> +                            if let Ok(consensus_scalar) =
> +                                ctx.automatic_conversion_consensus(&values)
> +                            {
> +                                ctx.convert_slice_to_common_leaf_scalar(
> +                                    &mut values,
> +                                    consensus_scalar,
> +                                )?;
> +                            }
> +
>                              let condition = self.expression(args.next()?, ctx)?;
>  
>                              args.finish()?;
>  
>                              ir::Expression::Select {
> -                                reject,
> -                                accept,
> +                                reject: values[0],
> +                                accept: values[1],
>                                  condition,
>                              }
>                          }
> ```

Just incorporated this into the patch, and dropped the refactors it made irrelevant. Thanks!

jimblandy reviewed (changes requested):
We shouldn't need to change the expression validator at all. Since the WGSL front end takes responsibility for applying automatic conversions and constant evaluation, the validator should never see an abstract value. I think that means that `proc::overloads::select` and its supporting code isn't needed either.

ErichDonGubler commented:
> We shouldn't need to change the expression validator at all. Since the WGSL front end takes responsibility for applying automatic conversions and constant evaluation, the validator should never see an abstract value. I think that means that `proc::overloads::select` and its supporting code isn't needed either.

Seems you're right! Stripped the majority of this PR, and force-pushed 817989b567a2.

ErichDonGubler committed:
refactor(wgsl-in): make `Scalar::concretize` `pub(in crate::front::wgsl)`

ErichDonGubler commented on naga/src/proc/constant_evaluator.rs:
```diff
@@ -2497,6 +2528,116 @@ impl<'a> ConstantEvaluator<'a> {
+
+        let select_single_component =
+            |this: &mut Self, reject_scalar, reject, accept, condition| {
+                let accept = this.cast(accept, reject_scalar, span)?;
```
Migrated from <https://github.com/gfx-rs/wgpu/pull/7602#discussion_r2067052453>: @jimblandy was wondering if we really needed `cast` here. So far as I can tell...yes.

ErichDonGubler replied:
Ah, another issue: We apparently coerce `0` and non-zero numeric values to `bool` with `cast`. That makes `select` accept strictly more things than it should in constant evaluation; is there a reason we _should_ allow such a cast?

ErichDonGubler replied:
@jimblandy: Poke on this.

ErichDonGubler replied:
Just posted this conversation as a full issue at <https://github.com/gfx-rs/wgpu/issues/7788>. See you there!

ErichDonGubler commented on naga/src/valid/expression.rs:
todo: Drop changes here. I'd originally changed this code, with a refactor that's no longer needed.

jimblandy commented on naga/tests/in/wgsl/select.wgsl:
```diff
@@ -0,0 +1,32 @@
+const_assert select(0xdeadbeef, 42f, false) == 0xdeadbeef;
+const_assert select(0xdeadbeefu, 42, false) == 0xdeadbeefu;
+const_assert select(0xdeadi, 42, false) == 0xdeadi;
+
+const_assert select(42f, 9001, true) == 9001;
+const_assert select(42f, 9001, true) == 9001f;
+const_assert select(42, 9001i, true) == 9001;
+const_assert select(42, 9001u, true) == 9001;
```
```suggestion
const_assert select(0xdeadbeef, 42f, false) == 0xdeadbeef;
const_assert select(0xdeadbeefu, 42, false) == 0xdeadbeefu;
const_assert select(0xdeadi, 42, false) == 0xdeadi;

const_assert select(42f, 0xdeadbeef, true) == 0xdeadbeef;
const_assert select(42, 0xdeadbeefu, true) == 0xdeadbeefu;
const_assert select(42, 0xdeadi, true) == 0xdeadi;

const_assert select(42f, 9001, true) == 9001;
const_assert select(42f, 9001, true) == 9001f;
const_assert select(42, 9001i, true) == 9001;
const_assert select(42, 9001u, true) == 9001;

const_assert select(9001, 42f, false) == 9001;
const_assert select(9001, 42f, false) == 9001f;
const_assert select(9001i, 42, false) == 9001;
const_assert select(9001u, 42, false) == 9001;
```

ErichDonGubler replied:
Committed as fb320b12.

jimblandy reviewed (commented):
Let's add this to `cts_runner/test.lst`:

```
webgpu:shader,execution,expression,call,builtin,select:*
```

With this PR, they all pass; without it, we get:

```
** Summary **
Passed  w/o warnings =  84 / 112 =  75.00%
Passed with warnings =   0 / 112 =   0.00%
Skipped              =   0 / 112 =   0.00%
Failed               =  28 / 112 =  25.00%
```

jimblandy commented on naga/src/front/wgsl/lower/mod.rs:
```diff
@@ -2483,12 +2484,70 @@ impl<'source, 'temp> Lowerer<'source, 'temp> {
+                            let diag_deets =
+                                |module: &ir::Module, ty: &ir::TypeInner, orig_expr| {
+                                    (
+                                        ctx.ast_expressions.get_span(orig_expr),
```
When we lower an AST expression to a Naga IR expression, we preserve its span. You can use `ctx.get_expression_span(lowered_expr)` to get these spans, so there's no need to retain the `foo_orig` handles.

ErichDonGubler replied:
I had originally authored this PR using `get_expression_span` on lowered expression handles, but I actually found gaps in the spans returned. For many runtime expressions in the tests I'm adding, we instead get `Span::default()` instead of the expected spans. 😐

I'd be happy to figure this particular issue out with you, but I suspect that it will need further work that's orthogonal to this PR. I don't think this should be blocking.

ErichDonGubler replied:
Implemented this change (outside this PR) and changed tests to demonstrate the regressions: c449dd01cdf8

ErichDonGubler replied:
Ah, this seems to be because the span lookup for locals appears to try to use something other than the span of the identifier referring to those locals; writing some quick debug diagnostics (fe6001684551) against this shader:

```wgsl
fn select_arrays(which: bool) -> i32 {
  var x: array<i32, 4>;
  var y: array<i32, 4> = array(1, 2, 3, 4);
  let s = select(x, y, which);
  return s[0];
}
```

…yields this output:

```
error: before
  ┌─ wgsl:4:18
  │
4 │   let s = select(x, y, which);
  │                  ^

error: after

error: before
  ┌─ wgsl:4:21
  │
4 │   let s = select(x, y, which);
  │                     ^

error: after

error: before
  ┌─ wgsl:4:24
  │
4 │   let s = select(x, y, which);
  │                        ^^^^^

error: after
  ┌─ wgsl:1:18
  │
1 │ fn select_arrays(which: bool) -> i32 {
  │                  ^^^^^

Could not parse WGSL:
error: unexpected argument type for `select` call
 = note: expected a scalar or a `vecN` of scalars
```

…where `before` and `after` diagnostics should be pointing to the same span, but they don't. If it's true that `get_expression_span` should be returning the same span as the expression used to get the `ir::Handle`, it looks like we have some nontrivial changes to make.

jimblandy replied:
This is a known limitation, filed as #7498. When we fix that, these diagnostics will magically get better.

You can work around it in your tests by making the operand more complex than a variable reference, like `y + 0` or something.

ErichDonGubler replied:
...but why would I accept bad diagnostics _for a user_ for something that's easy to migrate to once fixed? I'd strongly prefer that this code be migrated to `get_expression_span` _after_ it's fixed, because the diagnostics are okay _right now_. 😭

jimblandy replied:
Okay, that's fine.

jimblandy commented on naga/src/front/wgsl/lower/mod.rs:
```diff
@@ -2483,12 +2484,70 @@ impl<'source, 'temp> Lowerer<'source, 'temp> {
+                                |module: &ir::Module, ty: &ir::TypeInner, orig_expr| {
+                                    (
+                                        ctx.ast_expressions.get_span(orig_expr),
+                                        format!("`{:?}`", ty.for_debug(&module.types)),
```
Ideally we should hold on to the `TypeResolution` for these expressions, and use `ctx.as_diagnostic_display(resolution)` for generating diagnostics, as by the time we've reduced that to a `TypeInner`, we've lost the names of struct types.

ErichDonGubler replied:
Ah, this is because I was lacking familiarity with using the right types for `DiagnosticDisplay` functionality. I'd forgotten to make a task to come back and figure that out before claiming I was done. 

Pushed a `fixup!`: 09b3851b

ErichDonGubler commented:
@jimblandy: Added `webgpu:shader,execution,expression,call,builtin,select:*` to `cts_runner/test.lst` as a `fixup!`: b5fc75dd

Neat that this exists now! 😀

ErichDonGubler commented:
Marking this as ready for review, since the approach has been validated and it is in de-facto late stages of review.

jimblandy commented on naga/src/front/wgsl/lower/mod.rs:
```diff
@@ -2541,12 +2542,76 @@ impl<'source, 'temp> Lowerer<'source, 'temp> {
+                            ) -> &'a proc::TypeResolution {
+                                &ctx.typifier()[expr]
+                            }
+                            let diag_deets =
```
I don't want a code base full of names like this. Please rename to `diagnostic_details`.

ErichDonGubler replied:
Oh, dear, I forgot to change this, and had intended it to be a placeholder name written to keep my brain away from bikeshedding before finalizing logic. Thanks for calling me out!

ErichDonGubler replied:
Fixed as suggested with bc5421cf.

jimblandy commented on naga/src/front/wgsl/lower/mod.rs:
```diff
@@ -2541,12 +2542,76 @@ impl<'source, 'temp> Lowerer<'source, 'temp> {
 
                             args.finish()?;
 
+                            fn expr_ty_res<'a>(
+                                ctx: &'a ExpressionContext<'_, '_, '_>,
+                                expr: Handle<ir::Expression>,
+                            ) -> &'a proc::TypeResolution {
+                                &ctx.typifier()[expr]
+                            }
```
This function doesn't earn its keep. See notes below.

ErichDonGubler replied:
I agree; in previous versions of this PR, this was a more complicated body, but now it's trivial to as to not be worth it.

Fixed as suggested with 68a56cf4. LSP made this _trivial_ to inline. 😀

jimblandy commented on naga/src/front/wgsl/lower/mod.rs:
```diff
@@ -2541,12 +2542,76 @@ impl<'source, 'temp> Lowerer<'source, 'temp> {
+                            for (&value, orig_value) in
+                                values.iter().zip([reject_orig, accept_orig])
+                            {
+                                ctx.grow_types(value)?;
+
+                                let value_ty_res = expr_ty_res(ctx, value);
```
These two calls should be replaced with a use of the `resolve!` macro.

ErichDonGubler replied:
Fixed as suggested with a1a660e3.

jimblandy commented on naga/src/front/wgsl/lower/mod.rs:
```diff
@@ -2541,12 +2542,76 @@ impl<'source, 'temp> Lowerer<'source, 'temp> {
+                                    let [(reject_span, reject_type), (accept_span, accept_type)] =
+                                        [(reject_orig, reject), (accept_orig, accept)].map(
+                                            |(orig_expr, expr)| {
+                                                let ty_res = expr_ty_res(ctx, expr);
```
This should be:

```suggestion
                                                let ty_res = &ctx.typifier()[expr];
```

ErichDonGubler replied:
Fixed as suggested with 68a56cf4.

jimblandy reviewed (approved):
This is ready to land, with the following changes.

ErichDonGubler commented on cts_runner/test.lst:
```diff
@@ -18,6 +18,7 @@ webgpu:api,operation,rendering,depth:*
 webgpu:api,operation,rendering,draw:*
 webgpu:api,operation,shader_module,compilation_info:*
 webgpu:api,operation,uncapturederror:iff_uncaptured:*
+webgpu:shader,execution,expression,call,builtin,select:*
```
issue: Unfortunately, this test (`webgpu:shader,execution,expression,call,builtin,select:*`) doesn't pass in its entirety for us...yet. 😐 Looks like there are two sources of issues:

1. Some cases use built-ins in a `const` context, like `frexp`; from the Deno run on macOS just now:

	```
	  - EXCEPTION: Error: Unexpected validation error occurred: 
	    Shader '' parsing error: Not implemented as constant expression: Frexp built-in function
	       ┌─ wgsl:21:15
	       │
	    21 │     const f = frexp(abs(select(vec3(0.0, 1.0, 2.0), vec3(5.0, 6.0, 7.0), vec3(false, false, false))[0]));
	       │               ^^^^^ see msg
	    
	    : Not implemented as constant expression: Frexp built-in function
	        at DeviceHolder.attemptEndTestScope (file:///Users/runner/work/wgpu/wgpu/cts/out/webgpu/util/device_pool.js:457:13)
	        at eventLoopTick (ext:core/01_core.js:177:7)
	        at async DevicePool.release (file:///Users/runner/work/wgpu/wgpu/cts/out/webgpu/util/device_pool.js:101:7)
	        at async Promise.allSettled (<anonymous>)
	        at async AllFeaturesMaxLimitsGPUTestSubcaseBatchState.finalize (file:///Users/runner/work/wgpu/wgpu/cts/out/webgpu/gpu_test.js:153:21)
	        at async RunCaseSpecific.run (file:///Users/runner/work/wgpu/wgpu/cts/out/common/internal/test_group.js:696:9)
	        at async file:///Users/runner/work/wgpu/wgpu/cts/out/common/runtime/cmdline.js:214:5
	```

	This will be resolved by further progress on #4507.

2. `vec3` cases for `f16` that do _not_ `select` values using abstract types and `const` also run into #5262:

	```
	[fail] webgpu:shader,execution,expression,call,builtin,select:scalar:inputSource="uniform";component="h";overload="vec3" (74ms). Log:
	  - EXPECTATION FAILED: (vec3<f16>(0.0 (0x0000 subnormal), 1.0 (0x3c00), 2.0 (0x4000)), vec3<f16>(5.0 (0x4500), 6.0 (0x4600), 7.0 (0x4700)), true)
	        returned: vec3<f16>(0.0 (0x0000 subnormal), 1.0 (0x3c00), 2.0 (0x4000))
	        expected: vec3<f16>(5.0 (0x4500), 6.0 (0x4600), 7.0 (0x4700))
	    Error
	        at AllFeaturesMaxLimitsGPUTest.eventualAsyncExpectation (file:///Users/runner/work/wgpu/wgpu/cts/out/common/framework/fixture.js:254:24)
	        at AllFeaturesMaxLimitsGPUTest.expectGPUBufferValuesPassCheck (file:///Users/runner/work/wgpu/wgpu/cts/out/webgpu/gpu_test.js:694:10)
	        at file:///Users/runner/work/wgpu/wgpu/cts/out/webgpu/shader/execution/expression/expression.js:524:7
	        at processBatch (file:///Users/runner/work/wgpu/wgpu/cts/out/webgpu/shader/execution/expression/expression.js:415:5)
	        at eventLoopTick (ext:core/01_core.js:177:7)
	        at async Promise.all (index 0)
	        at async run (file:///Users/runner/work/wgpu/wgpu/cts/out/webgpu/shader/execution/expression/expression.js:446:3)
	        at async RunCaseSpecific.fn (file:///Users/runner/work/wgpu/wgpu/cts/out/webgpu/shader/execution/expression/call/builtin/select.spec.js:155:3)
	```

I suppose I could just get more granular paths, but that sounds very labor-intensive for something already getting checked in Firefox. I think it's better to wait until the related issues get resolved.

ErichDonGubler replied:
I'll comment this out with the relevant blockers noted.

jimblandy replied:
Agreed. We'll get to those.
